Documentation/ceph-best-practices.md

```diff
@@ -0,0 +1,516 @@
+
+Recommendations and calculations for OSD CPU are straightforward.
+
+Hardware recommendations:
+ * 1 x 2GHz CPU Thread per spinner
+ * 2 x GHz CPU Thread per SSD
+ * 4 x GHz CPU Thread per NVMe
+
+Examples:
+ * 8 HDDS journaled to SSD – `10 cores / 8 OSDs = 1.25 cores per OSD`
+ * 6 SSDs without journals – `12 cores / 6 OSDs = 2 cores per OSD`
+ * 8 SSDs journaled to NVMe – `20 cores / 8 OSDs = 2.5 cores per OSD`
+
+Note that resources are applied cluster-wide to all OSDs. If a cluster contains multiple OSD types,
+you must use the highest Requests for the whole cluster. For the examples below, a mixture of HDDs
+journaled to SSD and SSDs without journals would necessitate a Request for 2 cores.
+
+### Resource Requests - OSD RAM
+
+We have node hardware recommendations for OSD RAM usage, and we need to translate this to RAM
+requests on a per-OSD basis. The node-level recommendation we have is given below where
+`osd_memory_target`. This is a Ceph configuration described more afterwards.
+
+> `Total RAM required = [number of OSDs] x (1 GB + osd_memory_target) + 16 GB `
+
+Ceph OSDs will attempt to keep heap memory usage under a designated target size set via the
+`osd_memory_target` configuration option. Ceph’s default `osd_memory_target` is 4GB, and it is
+recommended to never decrease the `osd_memory_target` below 4GB. You may wish to increase this value
+to improve overall Ceph read performance by allowing the OSDs to use more RAM. While the total
+amount of heap memory mapped by the process should generally stay close to this target, there is no
+guarantee that the Linux kernel will actually reclaim memory that has been unmapped.
+
+As an example, for a node hosting 8 OSDs, memory Requests would be calculated as below.
+
+> `8 OSDs x (1GB + 4GB) + 16GB = 56GB per node`
+
+This gives us resource usage for each OSD.
+
+> `56GB / 8 OSDs = 7GB`
+
+Ceph has a feature that allows it to set `osd_memory_target` automatically when a Rook OSD Resource
+Request is set. However, Ceph sets this value 1:1 and does not leave overhead for waiting for the
+kernel to free memory. Therefore, we recommend setting `osd_memory_target` in Ceph explicitly, even
+if you wish to use the default value. Set Rook’s OSD resource requests accordingly and to a higher
+value than `osd_memory_target` by at least an additional 1GB. This is so Kubernetes does not
+schedule more applications or Ceph daemons onto a node than the node is likely to have RAM available for.
+
+OSD RAM Resource Requests come with the same cluster-wide Resource Requests note as for OSD CPU. Use
+the highest Requests for a cluster consisting of multiple different configurations of OSDs.
+
+### Resource Requests - Gateways
+
+For gateways, the best recommendation is to always tune your workload and daemon configurations.
+However, we do recommend the following initial configurations:
+
+**RGWs:**
+ * 6-8 CPU cores
+ * 64 GB RAM (32 GB minimum – may only apply to older "civetweb" protocol)
```

Comment:
Is the mention of civetweb really necessary? Because Rook only support Nautilus and higher "beast" will always be configured.

Reply:
This highly depends on the load so I'm not sure how we can even give a standard recommendation. An RGW would be just fine with 4GB if there are not many clients.